#70 generate streamfield block

nickmoreton opened the pull request.

nickmoreton committed:
add test to check shortcode custom HTML tags are promoted

shortcode handlers have a default is_top_level_html_tag = True

this can be overriden in developers own shortcode handlers

nickmoreton committed:
add custom_html_tag_prefix property to tests

update

nickmoreton committed:
include shortcode custom tags in bleach clean ALLOWED_TAGS

The beach filter needs to know about all shortcode custom HTML tags to prevent it escaping the <> characters

nickmoreton committed:
implement the shortcode detection into the block builder process

nickmoreton committed:
change image block fields

nickmoreton committed:
change image block template

nickmoreton committed:
change tests to a simpler fixture rather than passing in the raw_html.txt

This helps to pinpoint where any errors are occuring.

nickmoreton committed:
change construct_block method to output ImageBlock fields

nickmoreton committed:
update tests for latest block builder process

nickmoreton committed:
add test for devloper added shortcode handler

nickmoreton committed:
add extra class attrubites for custom HTML naming

nickmoreton committed:
validate shortcode handlers can only write valid streamfield blocks

The validation takes place in WordpressImporter class as the page model is required to validate the json for the body field

nickmoreton committed:
remove the need to register a shortcode with a name

nickmoreton committed:
update prefilter documentation

save

nickmoreton committed:
update shortcode handler documentation

nickmoreton committed:
change tests to check for type(int) rather than a specific value

there is a ticket to improve testing when fetching remote images: https://projects.torchbox.com/projects/wordpress-to-wagtail-importer-package/tickets/76

nimasmi commented on docs/prefilters.md:
```diff
@@ -1,125 +1,137 @@
 
+-- **filter_transform_shortcodes()** [source](wagtail-wordpress-import/wagtail_wordpress_import/prefilters/handle_shortcodes.py)
+
+This filter will parse the body content for Wordpress shortcodes and convert them to a custom HTML tag which is later converted to a `complex` StreamField block type. [See shortcode documentation](wagtail-wordpress-import/docs/shortcodes.md)
```
**Question**: Does 'complex' have a special meaning here? Why is it formatted as code?

nickmoreton replied:
I don't we even need to say that here, I will remove `complex`

nimasmi commented on docs/prefilters.md:
```diff
@@ -1,125 +1,137 @@
     },
+    {
+        "FUNCTION": "wagtail_wordpress_import.prefilters.transform_inline_styles",
+    },
```
**Issue**: Repetition. Did you mean `transform_shortcodes`?

nickmoreton replied:
Yes I did, will change it to `transform_shortcodes` but I will add it on line 130 as it needs to run there

nimasmi commented on docs/shortcodes.md:
````diff
@@ -1,140 +1,179 @@
-class FaqListShortcodeHandler(BlockShortcodeHandler):
-  shortcode_name = 'faqlist'
+```html
+<wagtail_block_card><img class="wp-image-46162 size-full" src="https://www.example.com/images/foo.jpg" />Card title</wagtail_block_caption>
````
**Issue**: Typo.

```suggestion
<wagtail_block_card><img class="wp-image-46162 size-full" src="https://www.example.com/images/foo.jpg" />Card title</wagtail_block_card>
```

nimasmi commented on docs/shortcodes.md:
````diff
@@ -1,140 +1,179 @@
 ```
 
-Your class should include `shortcode_name` as a class property with a value of the shortcode name that you need to transform.
+Your class should declare a class attribute of `shortcode_name` (str) with a value of the Wordpress shortcode name you need to transform. It should also declare a class attribute of `custom_html_tag_prefix` (str). In the example class above the class attributes will result in a custom HTML tag name of `wagtail_block_card`.
````
**Question**: Why did the custom tag prefix become configurable? Did I do that? Why is it a required parameter? Can we not use a default?

nickmoreton replied:
That was my idea :) I was thinking that the prefix might need to be different depending on how a developer might be parsing the soup in the construct_block() method. This was so they could overcome a use case we haven't thought of.

nimasmi replied:
> This was so they could overcome a use case we haven't thought of.

Okay. This is already a complicated feature, and [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) springs to mind. If we haven't thought of the use case, I don't think we need to write the code.

nickmoreton replied:
**Question** Would you prefer me to remove that ?

nimasmi replied:
Yes please. Is that drastic? I just think let's keep this as uncomplicated as we can get away with, at least until it's proven.

nickmoreton replied:
I don't think so. Will change it now.

nickmoreton replied:
I have removed the `custom_html_tag_prefix` from the CaptionHandler and BlockShortcodeHandler. Docs updated as well.

nimasmi commented on docs/shortcodes.md:
```diff
@@ -1,140 +1,179 @@
 
-For example: we need to create a shortcode handler to transform a `faqlist` Wordpress shortcode.
+For example: lets say we need to create a shortcode handler to transform a `card`  Wordpress shortcode.
+
```
**Suggestion**: Delete this paragraph.

```suggestion
```

nickmoreton replied:
Will do, you're right it's pretty obvious it's a contrived example.

nimasmi commented on docs/shortcodes.md:
```diff
@@ -1,140 +1,179 @@
-*The name you use in the **register decorator** must match the shortcode name you need to transform.*
-
-Your class will then perform the transformation of the Wordpress shortcode to a specific HTML tag with the name `wagtail_block_{shortcode_name}` during the import process.
+Decorate you class:
```
**Suggestion**: Typo.

```suggestion
Decorate your class:
```

nickmoreton replied:
Will fix

nimasmi commented on docs/shortcodes.md:
```diff
@@ -1,140 +1,179 @@
-  a custom HTML tag generated by pre_filter() 
+def construct_block(self, soup):
+  """Assuming the card has a `title` and `image` for its HTML content
+  As the `soup` parameter is a BeautifulSoup tag object you can use it's methods to extract the data you need.
```
**Suggestion** Fix typo.

```suggestion
  As the `soup` parameter is a BeautifulSoup tag object you can use its methods to extract the data you need.
```

nickmoreton replied:
Will fix

nimasmi commented on wagtail_wordpress_import/block_builder.py:
```diff
@@ -52,13 +59,21 @@ def get_builder_function(self, element):
+        conf_custom_tags = {}
+        for handler in SHORTCODE_HANDLERS:
+            cls = handler()
+            conf_custom_tags[cls.element_name] = handler
```
**Question**: Is this so Bleach doesn't later remove them? Or just so we can fetch the builder method from the Shortcode Handler class?

I'm still groking this, but do we run this `get_builder_function` method once per element in the soup. Could we farm out the looping over the SHORTCODE_HANDLERS? It seems repetitive.

nickmoreton replied:
> **Question**: Is this so Bleach doesn't later remove them? Or just so we can fetch the builder method from the Shortcode Handler class?

Yes bleach was escaping the opening < and closing > characters. Then the block builder failed to find the tag with BeautifulSoup but the ALLOWED_TAGS for bleach get updated in https://github.com/torchbox/wagtail-wordpress-import/blob/51aee21f5907626de0d5f7b8313568e6920cc4c3/wagtail_wordpress_import/prefilters/bleach_filter.py#L15

nickmoreton replied:
> I'm still groking this, but do we run this `get_builder_function` method once per element in the soup. Could we farm out the looping over the SHORTCODE_HANDLERS? It seems repetitive.

I will look again at this.

nickmoreton replied:
I think this could be done differently. We could change the register function to store a dict and have the ALLOWED_HTML_TAGS and the shortcode construct_block stored there when the shortcode handlers are loaded. This would mean we could store the functions as functions and then just call them I think. That would help with get_builder_function()

But as the ALLOWED_HTML_TAGS config is read every time we'd still need to append the shortcode element_name. It would be the same for the TAGS_TO_PROMOTE config here promote_child_tags()

nimasmi commented on wagtail_wordpress_import/block_builder.py:
```diff
@@ -84,14 +99,16 @@ def build(self):
+                if builder_function[0] == "html":
+                    self.blocks.append(builder_function[1](element))
+                elif builder_function[0] == "custom":
+                    self.blocks.append(builder_function[1]().construct_block(element))
```
**Comment**: This isn't terribly clear, though I see where it's come from. Do we need to store these as "html" and "custom"? Can we use something more readable, maybe an `isinstance` check instead, then we wouldn't need the tuple indexing?

nickmoreton replied:
I will look again at this.

nimasmi commented on wagtail_wordpress_import/importers/wordpress.py:
```diff
@@ -153,6 +161,21 @@ def run(self, *args, **kwargs):
+    @staticmethod
+    def check_stream_field_block_types(page, body):
+        """Body JSON is validated to check it is using only StreamField blocks declared in the model StreamField
+        An excpetion will stop the import process and display the ValueError
```
```suggestion
        An exception will stop the import process and display the ValueError
```

nickmoreton replied:
Will fix

nimasmi commented on wagtail_wordpress_import/prefilters/bleach_filter.py:
```diff
@@ -11,6 +12,13 @@ def filter_bleach_clean(html, options=None):
         CONF_ALLOWED_TAGS += options["ADDITIONAL_ALLOWED_TAGS"]
 
+    # Registered shortcode handlers generate custom tags
+    # so they need to be added too ALLOWED_TAGS
```
```suggestion
    # so they need to be added to ALLOWED_TAGS
```

nickmoreton replied:
Will fix

nimasmi commented on wagtail_wordpress_import/test/tests/test_block_builder.py:
```diff
@@ -199,16 +226,24 @@ def test_images_linked_rich_text(self):
+
+        value_soup = BeautifulSoup(self.blocks[0]["value"], "html.parser")
+        # embed_id is an image id. it should be returned as an integer but we cannot
+        # reply on the value returned here so check if it is an integer.
```
```suggestion
        # rely on the value returned here so check if it is an integer.
```

nickmoreton replied:
Will fix

nimasmi commented on wagtail_wordpress_import/test/tests/test_block_builder.py:
```diff
@@ -199,16 +226,24 @@ def test_images_linked_rich_text(self):
+        # there is a ticket to improve testing when fetching remote images:
+        # https://projects.torchbox.com/projects/wordpress-to-wagtail-importer-package/tickets/76
+        embed_id = int(value_soup.find("embed").attrs["id"])
+        self.assertIsInstance(embed_id, int)
```
**Question**: What circumstances would cause this assertion to fail?

```python
value = int(foo)
self.assertIsInstance(value, int)
```

nickmoreton replied:
Actually it would fail in line 245 if it wasn't a valid string that converted to an int. I'll rethink it. It came about that those tests failed in the CI when I pushed the branch but only when it ran tests with postgres db. Do you know why that would happen when sqlite db all passed OK?

nimasmi replied:
Sadly, no. Was it previously just `self.assertIsInstance(embed_id, int)`? I see the failing test at https://github.com/torchbox/wagtail-wordpress-import/runs/4136707582?check_suite_focus=true, but that's a different integer, not a type discrepancy. Do you have a link to the offending test run?

nickmoreton replied:
Yes thats the failing test, the integers being returned were not the same as they were on sqlite. My local dev uses postgres but I wasn't seeing any errors locally with that DB

nimasmi replied:
Ah, I see, thanks.

Having read this test case, I'm uncertain why the `<img>` HTML tag ends up as an `<embed>` tag in the final HTML, but if we're to test that the ID matches whatever new image is imported, we could use a query. A speculative example:

```python
image = Image.objects.get(file__name="bruno-4-runner.jpg")  # based on `block_builder_defaults.get_or_save_image`
self.assertEqual(int(embed_id), image.id)
```

nimasmi replied:
Since this PR was merged without the above change, I've opened ticket https://projects.torchbox.com/projects/wordpress-to-wagtail-importer-package/tickets/86 to address it.

nimasmi commented on wagtail_wordpress_import/test/tests/test_shortcode_handlers.py:
```diff
@@ -236,62 +248,59 @@ def test_included_shortcodes(self):
+        wagtail_custom_html = BeautifulSoup(
+            '<wagtail_block_caption id="46162" align="aligncenter" width="600">'
+            '<img class="size-full" '
+            'src="https://www.budgetsaresexy.com/images/j-money-family-portrait.jpg" '
```
**Issue**: We have a remote URL here. Are we querying this during the tests?

nickmoreton replied:
Yes that gets queried here: https://github.com/torchbox/wagtail-wordpress-import/blob/77ad1c0aa3a1ec41c829bb2d5b89631f97727391/wagtail_wordpress_import/block_builder_defaults.py#L222 

It's a function that's also used when linking rich_text images.

We have a ticket to make this better in testing: https://projects.torchbox.com/projects/wordpress-to-wagtail-importer-package/tickets/76

nickmoreton committed:
code review documentation changes

nickmoreton committed:
code review typo's

nickmoreton committed:
remove the custom_html_tag_prefix attribute

Removed from BlockShortcodeHandler, CaptionHandler and updated the docs to reflect the change.

nickmoreton committed:
refactor the block builder conf

nickmoreton committed:
Change the block builder config

This restructures the config list to a dict. I think earlier on the intentionwas to be able to pass extra options to the functions but we haven't done that anywhere.

nickmoreton committed:
change tests for block builder conf

I have also altered test_images_linked_rich_text to check if a value is returned rather than a specific image id

nickmoreton committed:
update fixtures and block builder tests

I fell foul of my own fixture warning. But I don't know why this happend after a rebase.

nimasmi reviewed (approved):
If the latest discussion on the broken test case makes sense, then please address it. Otherwise, these changes look good.
